### PR TITLE
Form

### PR DESCRIPTION
**Pull request by ewapotorecka**
- added formik and yup 
- added validation to habit form
- added first tests for form

### CODE REVIEW

File: src/components/Pages/Form/Form.tsx [hunk 1]
```diff
@@ -17,62 +17,84 @@ import { setNewHabit } from "../../../app/features/habit/habitSlice";
...
+
   const dispatch = useDispatch();
   const navigate = useNavigate();
+  const formik = useFormik({
+    initialValues: {
+      goal: "",
+      schema: "",
+      rewards: [],
+    },
+    validationSchema: validationSchema,
```
Comment (ma2ciek) [inline src/components/Pages/Form/Form.tsx:40]:
```suggestion
    validationSchema,
```

File: src/components/Pages/Form/Form.tsx [hunk 2]
```diff
@@ -17,62 +17,84 @@ import { setNewHabit } from "../../../app/features/habit/habitSlice";
...
-              id="goal-input"
-              aria-describedby="goal-input-helper-text"
-              value={goalInput}
-              onChange={(e) => setGoalInput(e.target.value)}
+              id="goal"
+              name="goal"
+              aria-describedby="goal-helper-text"
+              value={formik.values.goal}
+              onChange={formik.handleChange}
+              error={formik.touched.goal && Boolean(formik.errors.goal)}
```
Comment (ma2ciek) [inline src/components/Pages/Form/Form.tsx:71]: what is the type of `formik.errors.goal`?

File: src/components/Pages/Form/Form.tsx [hunk 3]
```diff
@@ -127,22 +155,27 @@ const HabitForm = () => {
...
-                onDelete={() =>
-                  setRewards(
-                    rewards.filter(
-                      (rewardToDelete) => rewardToDelete.id !== reward.id
-                    )
-                  )
-                }
-              />
-            ))}
+            {formik.values.rewards.length > 0 &&
```
Comment (ma2ciek) [inline src/components/Pages/Form/Form.tsx:158]: not needed

File: src/components/Pages/Form/Form.tsx [hunk 4]
```diff
@@ -127,22 +155,27 @@ const HabitForm = () => {
...
+              formik.values.rewards.map(
+                (reward: { label: string; id: number }) => (
+                  <Chip
+                    label={reward.label}
+                    key={reward.id}
+                    onDelete={() =>
+                      formik.setFieldValue(
+                        "rewards",
+                        formik.values.rewards.filter(
+                          (rewardToDelete: { id: number; label: string }) =>
```
Comment (ma2ciek) [inline src/components/Pages/Form/Form.tsx:168]: The typing should not be needed here. I hope that `formik.values.rewards` is already typed. Or isn't it?

File: src/tests/Form.test.tsx
Comment (ma2ciek) [inline src/tests/Form.test.tsx:1]: Looks great! 👌🎉

File: src/tests/Form.test.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    </Provider>
+  );
+
+  const input: HTMLInputElement = screen.getByTestId("reward-input");
+  const addButton = screen.getByTestId("add-reward");
+
+  userEvent.type(input, "new reward");
+  userEvent.click(addButton);
+
+  expect(screen.getByText(/new reward/i)).toBeInTheDocument();
```
Comment (ma2ciek) [inline src/tests/Form.test.tsx:41]: let's be more precise here where to search for that string.

File: src/tests/Form.test.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  const submit = screen.getByTestId("create-habit");
+
+  fireEvent.click(submit);
+
+  expect(await screen.findByText(/Goal is required/i)).toBeInTheDocument();
+  expect(await screen.findByText(/Schema is required/i)).toBeInTheDocument();
+  expect(
+    await screen.findByText(/Add at least one reward/i)
+  ).toBeInTheDocument();
```
Comment (ma2ciek) [inline src/tests/Form.test.tsx:61]: I'd be also a bit more precise here in this assertions.

File: src/tests/Form.test.tsx (hunk 5 above)
Comment (ewapotorecka) [reply src/tests/Form.test.tsx:41]: so here you would suggest looking for this string it the container displaying rewards?

File: src/components/Pages/Form/Form.tsx [hunk 7]
```diff
@@ -143,7 +144,7 @@ const HabitForm = () => {
               }
             />
             <FormHelperText id="rewards-helper-text">
-              {formik.touched.rewards && formik.errors.rewards}
+              {formik.touched.rewards && (formik.errors.rewards as string)}
```
Comment (ma2ciek) [inline src/components/Pages/Form/Form.tsx:148]: Do you need to type case here?

File: src/tests/Form.test.tsx (hunk 5 above)
Comment (ma2ciek) [reply src/tests/Form.test.tsx:41]: yes

File: src/components/Pages/Form/Form.tsx (hunk 7 above)
Comment (ewapotorecka) [reply src/components/Pages/Form/Form.tsx:148]: I think so

**Review by ma2ciek — APPROVED** [review]

File: src/components/Pages/Form/Form.tsx (hunk 7 above)
Comment (ma2ciek) [reply src/components/Pages/Form/Form.tsx:150]: So what is the default type here `formik.errors.rewards`? `undefined` | string?

**Pull request merged**